protocol/grpc/message.go

```diff
@@ -0,0 +1,211 @@
+)
+
+const (
+	prefix      = "ce-"
+	contenttype = "contenttype"
+	dataSchema  = "dataschema"
+	subject     = "subject"
+	time        = "time"
+)
+
+var specs = spec.WithPrefix(prefix)
+
+// Message represents a gRPC message.
+// This message *can* be read several times safely
+type Message struct {
+	internal *pb.CloudEvent
+	version  spec.Version
+	format   format.Format
+}
+
+// Check if Message implements binding.Message
+var (
+	_ binding.Message               = (*Message)(nil)
+	_ binding.MessageMetadataReader = (*Message)(nil)
+)
+
+func NewMessage(msg *pb.CloudEvent) *Message {
+	var f format.Format
+	var v spec.Version
+	if msg.Attributes != nil {
+		if contentType, ok := msg.Attributes[contenttype]; ok && format.IsFormat(contentType.GetCeString()) {
+			f = format.Lookup(contentType.GetCeString())
+		} else if s := msg.SpecVersion; s != "" {
+			v = specs.Version(s)
+		}
+	}
+	return &Message{
+		internal: msg,
+		version:  v,
+		format:   f,
+	}
+}
+
+func (m *Message) ReadEncoding() binding.Encoding {
+	if m.version != nil {
+		return binding.EncodingBinary
+	}
+	if m.format != nil {
+		return binding.EncodingStructured
+	}
+
+	return binding.EncodingUnknown
+}
+
+func (m *Message) ReadStructured(ctx context.Context, encoder binding.StructuredWriter) error {
+	if m.version != nil {
+		return binding.ErrNotStructured
+	}
+	if m.format == nil {
+		return binding.ErrNotStructured
+	}
+
+	return encoder.SetStructuredEvent(ctx, m.format, bytes.NewReader(m.internal.GetBinaryData()))
+}
+
+func (m *Message) ReadBinary(ctx context.Context, encoder binding.BinaryWriter) error {
+	if m.format != nil {
+		return binding.ErrNotBinary
+	}
+
+	if m.internal.SpecVersion == "" {
+		err := encoder.SetAttribute(m.version.AttributeFromKind(spec.SpecVersion), m.internal.SpecVersion)
+		if err != nil {
+			return err
+		}
+	}
+	if m.internal.Id != "" {
+		err := encoder.SetAttribute(m.version.AttributeFromKind(spec.ID), m.internal.Id)
+		if err != nil {
+			return err
+		}
+	}
+	if m.internal.Source != "" {
+		err := encoder.SetAttribute(m.version.AttributeFromKind(spec.Source), m.internal.Source)
+		if err != nil {
+			return err
+		}
+	}
+	if m.internal.Type != "" {
+		err := encoder.SetAttribute(m.version.AttributeFromKind(spec.Type), m.internal.Type)
+		if err != nil {
+			return err
+		}
+	}
+
+	for name, value := range m.internal.Attributes {
+		v, err := valueFrom(value)
+		if err != nil {
+			return fmt.Errorf("failed to convert attribute %s: %s", name, err)
+		}
+
+		if strings.HasPrefix(name, prefix) {
+			attr := m.version.Attribute(name)
+			if attr != nil {
+				switch attr.Kind() {
+				case spec.DataContentType, spec.DataSchema, spec.Subject, spec.Time:
+					err = encoder.SetAttribute(attr, v)
+					if err != nil {
+						return err
+					}
+				}
+			} else {
+				err = encoder.SetExtension(strings.TrimPrefix(name, prefix), v)
+				if err != nil {
+					return err
+				}
+			}
+		} else if name == contenttype {
```

Comment:
what if there's both 'contenttype' and 'ce-datacontenttype' ?

Reply:
'ce-datacontenttype' will be handled by: https://github.com/cloudevents/sdk-go/blob/ab0722849ab42679f5dee3f08eb85cd6c5e7078f/protocol/grpc/message.go#L122-L128

Reply:
yes but which one wins? Should we require them to be the same (if both are there)?

Reply:
When both 'ce-datacontenttype' and 'contenttype' are set, 'ce-datacontenttype' takes precedence. In practical scenarios, only 'contenttype' is necessary. Check how the 'contenttype' attribute is configured here:
https://github.com/cloudevents/sdk-go/blob/65239604c031066dab2a25bde0719c898566c3db/protocol/grpc/write_message.go#L121-L130

This follow the similar approach that the Kafka protocol and PubSub protocol take, refer to the implementations in the CloudEvents SDK for Kafka (https://github.com/cloudevents/sdk-go/blob/main/protocol/kafka_sarama/v2/write_producer_message.go#L103-L114) and PubSub (https://github.com/cloudevents/sdk-go/blob/main/protocol/pubsub/v2/write_pubsub_message.go#L70-L80).